bot: Add a bot owner tab in realm_activity page.

edith007 opened the pull request:
This PR aims to solve the issue https://github.com/zulip/zulip/issues/18275

Add a bot owner tab in realm_activity page to know who created the bot on the realm, in case they're misbehaving.

Screenshot: 

![image](https://user-images.githubusercontent.com/53316982/117871002-7fa34600-b2ba-11eb-8186-531528704b30.png)

![image](https://user-images.githubusercontent.com/53316982/117871032-8e89f880-b2ba-11eb-8c29-607312ecad1a.png)

timabbott commented on analytics/views.py:
```diff
@@ -1728,6 +1730,9 @@ def by_used_time(row: Dict[str, Any]) -> str:
     ]
 
+    if is_bot:
+        cols.insert(2, "Bot Owner")
```
We probably want a `role` column for users who are not a bot owner; maybe worth fixing that while you're working on this?

alexmv commented:
For my use case, in particular, the _name_ of the bot owner is not as interesting as their _email_.  Maybe include both, a la `name <email@example.com>` ?

timabbott commented:
Yeah, or if we think space is tight, `name` with a mailto: link for the `delivery_email`.

timabbott commented on zerver/tests/test_presence.py:
```diff
@@ -59,14 +59,14 @@ def test_activity(self, unused_mock: mock.Mock) -> None:
             self.assertEqual(result.status_code, 200)
 
-        self.assert_length(queries, 4)
+        self.assert_length(queries, 5)
```
I suspect you can not need any new changes if you change this:
`    records = records.select_related("user_profile", "client").only(*fields)`
to
`    records = records.select_related("user_profile__bot_owner", "client").only(*fields)`

Can you try that?

edith007 replied:
> reduced the number of queries from 14 to 10. Earlier the total number of queries were 8 but these additional two queries are to fetch the bot owner.

I made the change and the number of queries reduced from 10 to 8, but adding role column in human tab causes the query increase by 1.

timabbott commented on analytics/views.py:
```diff
@@ -1728,6 +1754,9 @@ def by_used_time(row: Dict[str, Any]) -> str:
     ]
 
+    if is_bot:
+        cols[2] = "Bot Owner"
```
it might be better to have the column be `"Role" if not is_bot else "Bot owner"`.

alexmv replied:
Tangential, but the last 4 columns are not really useful on the bots page.  It'd be nice to remove them, possibly as a follow-up diff.

timabbott commented:
Posted a few comments, but otherwise this lgtm.

edith007 commented:
Hi @timabbott @alexmv I have addressed all your comment and the PR is ready for review.

edith007 committed:
realm_activity: Add Role and Bot Owner tabs.

Add "Role" to the Humans tab and "Bot Owner" to the Bots tab of
realm_activity page.

Fixes: zulip#18275

edith007 commented:
Hi @timabbott! @alexmv! 
This PR is ready for review!

alexmv commented on analytics/views/realm_activity.py:
```diff
@@ -71,6 +81,12 @@ def is_recent(val: Optional[datetime]) -> bool:
         sent_count = get_count(user_summary, "send")
         cells = [user_summary["name"], email_link, sent_count]
+        role = user_summary["role"]
+        cells = [user_summary["name"], email_link, role, sent_count]
+
+        if "bot_owner" in user_summary:
+            cells[2] = get_bot_owner_email(user_summary)
```
nit -- I'd inline this logic:
```suggestion
        cells = [
            user_summary["name"],
            email_link,
            role if "bot_owner" not in user_summary else get_bot_owner_email(user_summary),
            sent_count,
        ]

alexmv commented on analytics/views/realm_activity.py:
```diff
@@ -62,6 +66,12 @@ def get_count(user_summary: Dict[str, Dict[str, str]], k: str) -> str:
+    def get_bot_owner_email(user_summary: Dict[str, Dict[str, str]]) -> mark_safe:
+        email = user_summary["bot_owner_email"]
+        bot_owner = user_summary["bot_owner"]
+        email_link = f'<a href="mailto:{email}">{bot_owner}</a>'
+        return mark_safe(email_link)
```
Is this really "safe"?  That is, if my owner name is `</a><a href="script:alert('!')">Alex`, what happens?

I suspect we need to escape these before we insert them.  See [`format_html` instead of `mark_safe`](https://docs.djangoproject.com/en/3.2/ref/utils/#django.utils.html.format_html).

alexmv commented on analytics/tests/test_activity_views.py:
```diff
@@ -51,4 +51,4 @@ def test_activity(self, unused_mock: mock.Mock) -> None:
             self.assertEqual(result.status_code, 200)
 
-        self.assert_length(queries, 4)
+        self.assert_length(queries, 5)
```
This points to needing to adjust the fields and prefetch in `get_user_activity_records_for_email` in `analytics/views/user_activity.py` like you adjusted them in `get_user_activity_records_for_realm` in `analytics/views/realm_activity.py`.  IF you do that, this shouldn't need a change.

It's less critical for this page, because it's not O(n), but it'd be nice to do.

alexmv reviewed (commented):
Thanks for coming back to this!  Some comments inline -- a few very minor nits, as well as what I think is an important escaping bug to fix to prevent XSS.
